Minifi 227

benqiu2016 opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

phrocker commented on libminifi/src/Site2SiteClientProtocol.cpp:
```diff
@@ -682,6 +682,7 @@ bool Site2SiteClientProtocol::receive(std::string transactionID,
 
 bool Site2SiteClientProtocol::send(
     std::string transactionID, DataPacket *packet, std::shared_ptr<FlowFileRecord> flowFile,
+	uint8_t *payload, int length,
```
Can payload be embedded into DataPacket, thus  simplifying this function?

benqiu2016 replied:
send can take flowfile or raw bytes, DataPacket is more like a hold for the meta info like transaction, etc.

phrocker replied:
As someone reading this that does not make sense. Why not make DataPacket hold the "packet"?

benqiu2016 replied:
OK. will change that.

phrocker commented on libminifi/src/provenance/ProvenanceRepository.cpp:
```diff
@@ -36,6 +36,7 @@ void ProvenanceRepository::run() {
     uint64_t curTime = getTimeMillis();
     uint64_t size = repoSize();
     if (size >= purgeThreshold) {
+      std::lock_guard<std::mutex> lock(mutex_);
```
What makes this necessary? is there a way to avoid locking? LevelDB is thread safe due to process level locking.

benqiu2016 replied:
yes, levelDB is thread save, but when we purge the record in provenanceRepository task, i do not want to ProvenanceReportTask to pick the record and ship to NiFi.

phrocker replied:
If LevelDB is thread safe and is being picked up that means at that moment in the thread which picks it it is not eligible to be purged and is eligible to be sent. That means you can atomically remove said item and thread A will never get it to send OR lock all access. If you need locking here above LevelDB we should change the access pattern.

benqiu2016 replied:
OK

phrocker commented on libminifi/src/provenance/ProvenanceTaskReport.cpp:
```diff
@@ -0,0 +1,221 @@
+
+	if (context->getProperty(batchSize.getName(), value) && core::Property::StringToInt(value, lvalue))
+	{
+		batch = (int) lvalue;
```
why is (int) necessary? Why not use the same type as lvalue?

phrocker commented on libminifi/src/provenance/ProvenanceTaskReport.cpp:
```diff
@@ -0,0 +1,221 @@
+		recordJson["transitUri"] = record->getTransitUri().c_str();
+		recordJson["remoteIdentifier"] = record->getSourceSystemFlowFileIdentifier().c_str();
+		recordJson["alternateIdentifier"] = record->getAlternateIdentifierUri().c_str();
+		recordJson["application"] = "MiNiFi Flow";
```
Maybe application name should be externalized?

phrocker commented on libminifi/include/provenance/ProvenanceTaskReport.h:
```diff
@@ -0,0 +1,88 @@
+namespace provenance {
+
+//! ProvenanceTaskReport Class
+class ProvenanceTaskReport: public core::Processor {
```
Would a more applicable name be ProvenanceTaskReporter?

apiri replied:
would also be up for a rename, preferably SiteToSiteProvenanceReportingTask

would also like to see this namespaced to org.apache.nifi.reporting

phrocker commented on libminifi/src/provenance/ProvenanceTaskReport.cpp:
```diff
@@ -0,0 +1,221 @@
+
+	returnProtocol(std::move(protocol_));
+
+	return;
```
this isn't necessary.

benqiu2016 replied:
i want to return the same so that it can be reused in next trigger.

phrocker replied:
I'm sorry I don't follow. an empty return on a void isn't necessary. It's unnecessary code.

benqiu2016 replied:
sorry, i will remove the return. i thought you mean returnProtocol.

benqiu2016 commented on libminifi/src/provenance/ProvenanceTaskReport.cpp:
```diff
@@ -0,0 +1,221 @@
+	relationships.insert(relation);
+	setSupportedRelationships(relationships);
+}
+
```
we may want to add specific S2S protocol logic for shipping provenance report data. It make the code more flexible and decoupled.

phrocker replied:
I don't follow. How does duplicating code decouple and add flexibility?

benqiu2016 replied:
right now, we propose the Provenance Report property the same as RemoteProcess Group as hostName and Port, it may change. Also the way s2s protocol negotiation may change. For example. Remote Process Group may support round robin between different NiFi server instead of hardcode the hostName and port in the processor property. The decouple of these two provide more flexible and it make sense because they behave differently.

phrocker replied:
I don't see how potential change obviates the need to minimize duplicate code. 

If I may be able to rephrase your response, am I correct in understanding that you mean: "I am copying and pasting code into a class because it may eventually change. Therefore, I don't wish to encapsulate this functionality elsewhere. As a result, change may make that copied code be different at some point?" 

I want to make sure I understand you before I can respond.

benqiu2016 replied:
what i mean is that they are different processor and it make sense to make the code path differently. Certainly we can extract that returnProtocol and getNextProtocol to a parent class that both ProvenanceTaskReport and RemoteProcessGroup can inherit from if it is what you mean by duplicate code.

benqiu2016 commented on thirdparty/jsoncpp/devtools/batchbuild.py:
```diff
@@ -0,0 +1,278 @@
+from __future__ import print_function
```
the tests source code are there because it was from the drop for jsoncpp. but i disable build and run the same in the minifi build for the jasoncpp cmake.

phrocker replied:
Remove the code if it is unnecessary.

benqiu2016 replied:
OK.

benqiu2016 commented on libminifi/src/Site2SiteClientProtocol.cpp:
```diff
@@ -1240,6 +1257,88 @@ void Site2SiteClientProtocol::transferFlowFiles(
   return;
 }
 
+void Site2SiteClientProtocol::transferBytes(core::ProcessContext *context, core::ProcessSession *session, uint8_t *payload, int length,
```
but we need to copy the payload to vector<uint8_t>, it is overhead

phrocker replied:
No @benqiu2016 that is why std::move(...) exists.

benqiu2016 replied:
Can we move a byte array to vector<uint8_t>?

phrocker replied:
I can see the confusion. I was referring to a desire to minimize pointer usage and more strongly type functions within a public access. 

This we are trying to avoid: 
  *Having arbitrarily large sized arrays. payload comes in with a ptr and length comes in as a value well above what the NiFi instance could support. 
  *What happens to NiFI if length is accidentally negative? You convert a -1 to uint64_t meaning that we would attempt to send a value well outside of our process control. 
  * You have a conditional in send in transferBytes, both of which are public. what would happen if someone uses the protocol to call send instead of transferBytes with a payload and non null flowFile? The payload would be ignored. This is not ideal since send/transferbytes are don't differentiate each other very well.
  
What inherent need is there to have a generic function that has few safeguards? Further, this function was intended to be used for sending strings. Why not use std::string instead of uint8t*,int ? your calls write the size with length without any checks, whereas the usage of std::vector OR std::string make this unnecessary and allow us to be more defensive.

Why aren't you using writeUTF ? Do you have the link to the respective java code for the serialization?

benqiu2016 replied:
how can i do a std::move from a string to a std::vector<uint8_t> without copy?
 http://stackoverflow.com/questions/10445042/stdmove-between-stdstring-and-stdvectorunsigned-char
pass the raw byte looks OK to me and make the code much simple in this case instead of convert from one type to another.

benqiu2016 replied:
OK. will change to pass std::string.

phrocker commented:
Hey @benqiu2016 is this ready to continue to be reviewed? I see the build did not succeed in travis.

benqiu2016 commented:
@phrocker  
97%] Building C object thirdparty/civetweb-1.9.1/src/CMakeFiles/c-executable.dir/main.c.o
Linking C executable civetweb
/usr/include/x86_64-linux-gnu/bits/stdio2.h: In function ‘send_file_data’:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:290:2: error: call to ‘__fread_chk_warn’ declared with attribute warning: fread called with bigger size * nmemb than length of destination buffer [-Werror]
  return __fread_chk (__ptr, __bos0 (__ptr), __size, __n, __stream);
  ^
lto1: all warnings being treated as errors
lto-wrapper: /usr/bin/gcc returned 1 exit status
/usr/bin/ld: lto-wrapper failed
collect2: error: ld returned 1 exit status
make[2]: *** [thirdparty/civetweb-1.9.1/src/civetweb] Error 1
make[1]: *** [thirdparty/civetweb-1.9.1/src/CMakeFiles/c-executable.dir/all] Error 2
make: *** [all] Error 2

phrocker commented:
Hey @benqiu2016, the master build (https://travis-ci.org/apache/nifi-minifi-cpp/builds/219032189 ) works for both Linux and Mac.  I'm uncertain what you meant by pasting the error from your Travis output, but I'll be happy to take a stab. It looks like you pasted the Linux output, which fails because of build properties in Linux. My guess this is is in your CMake, perhaps, but I can't be certain without pulling it down. I'll take a look at that tomorrow as I'm currently running a Linux build and it succeeds from master. 

The Mac build failed due to linter errors, which was a recent addition. This is easy to solve by typing make linter and resolving the output. Let me know if I can help with that.

benqiu2016 commented:
my MAC build linter is OKW12612:build binqiu$ make linter
/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//Connection.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ConfigurableComponent.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ConfigurationFactory.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Connectable.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Core.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/FlowConfiguration.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/FlowFile.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/logging/BaseLogger.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/logging/LogAppenders.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/logging/Logger.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ProcessContext.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ProcessGroup.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Processor.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ProcessorConfig.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ProcessorNode.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ProcessSession.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/ProcessSessionFactory.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Property.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Relationship.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/repository/FlowFileRepository.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Repository.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/RepositoryFactory.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/Scheduling.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//core/yaml/YamlConfiguration.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//EventDrivenSchedulingAgent.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//Exception.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//FlowController.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//FlowControlProtocol.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//FlowFileRecord.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/BaseStream.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/ClientSocket.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/CRCStream.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/DataStream.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/EndianCheck.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/Serializable.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/Sockets.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/StreamFactory.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/tls/TLSSocket.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//io/validation.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/AppendHostInfo.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/ExecuteProcess.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/GenerateFlowFile.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/GetFile.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/ListenHTTP.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/ListenSyslog.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/LogAttribute.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/PutFile.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//processors/TailFile.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//properties/Configure.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//provenance/Provenance.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//provenance/ProvenanceRepository.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//provenance/ProvenanceTaskReport.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//RemoteProcessorGroupPort.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//ResourceClaim.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//SchedulingAgent.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//Site2SiteClientProtocol.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//Site2SitePeer.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//ThreadedSchedulingAgent.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//TimerDrivenSchedulingAgent.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//utils/FailurePolicy.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//utils/StringUtils.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//utils/ThreadPool.h,/Users/binqiu/report/nifi-minifi-cpp/libminifi/include//utils/TimeUtil.h,
/Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Configure.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Connection.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ConfigurableComponent.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ConfigurationFactory.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Connectable.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Core.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/FlowConfiguration.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/FlowFile.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/logging/BaseLogger.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/logging/LogAppenders.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/logging/Logger.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessGroup.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Processor.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessorNode.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessSession.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessSessionFactory.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Property.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/repository/FlowFileRepository.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Repository.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/RepositoryFactory.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/yaml/YamlConfiguration.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//EventDrivenSchedulingAgent.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//FlowController.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//FlowControlProtocol.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//FlowFileRecord.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/BaseStream.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/ClientSocket.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/CRCStream.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/DataStream.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/EndianCheck.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/Serializable.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/StreamFactory.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/tls/TLSSocket.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/AppendHostInfo.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/ExecuteProcess.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/GenerateFlowFile.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/GetFile.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/ListenHTTP.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/ListenSyslog.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/LogAttribute.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/PutFile.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/TailFile.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//provenance/Provenance.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//provenance/ProvenanceRepository.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//provenance/ProvenanceTaskReport.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//RemoteProcessorGroupPort.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//ResourceClaim.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//SchedulingAgent.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Site2SiteClientProtocol.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Site2SitePeer.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//ThreadedSchedulingAgent.cpp /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//TimerDrivenSchedulingAgent.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Configure.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Connection.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ConfigurableComponent.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ConfigurationFactory.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Connectable.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Core.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/FlowConfiguration.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/FlowFile.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/logging/BaseLogger.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/logging/LogAppenders.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/logging/Logger.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessGroup.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Processor.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessorNode.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessSession.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/ProcessSessionFactory.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Property.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/repository/FlowFileRepository.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/Repository.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/RepositoryFactory.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//core/yaml/YamlConfiguration.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//EventDrivenSchedulingAgent.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//FlowController.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//FlowControlProtocol.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//FlowFileRecord.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/BaseStream.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/ClientSocket.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/CRCStream.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/DataStream.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/EndianCheck.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/Serializable.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/StreamFactory.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//io/tls/TLSSocket.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/AppendHostInfo.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/ExecuteProcess.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/GenerateFlowFile.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/GetFile.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/ListenHTTP.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/ListenSyslog.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/LogAttribute.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/PutFile.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//processors/TailFile.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//provenance/Provenance.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//provenance/ProvenanceRepository.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//provenance/ProvenanceTaskReport.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//RemoteProcessorGroupPort.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//ResourceClaim.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//SchedulingAgent.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Site2SiteClientProtocol.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//Site2SitePeer.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//ThreadedSchedulingAgent.cpp
Done processing /Users/binqiu/report/nifi-minifi-cpp/libminifi/src//TimerDrivenSchedulingAgent.cpp
Total errors found: 0
Built target linter
HW12612:build binqiu$

benqiu2016 commented:
@phrocker it looks like the travis for MAC is OK.

apiri commented on README.md:
```diff
@@ -53,6 +53,7 @@ Perspectives of the role of MiNiFi should be from the perspective of the agent a
   * ListenSyslog
   * ExecuteProcess
   * AppendHostInfo
+  * ProvenanceTaskReport
```
We should make this SiteToSiteProvenanceReportingTask for consistency between instance types.

apiri commented on README.md:
```diff
@@ -286,6 +287,26 @@ Additionally, users can utilize the MiNiFi Toolkit Converter (version 0.0.1 - sc
+ 
+    - name: ProvenanceTaskReport
+      id: 471deef6-2a6e-4a7d-912a-81cc17e3a209
+      class: org.apache.nifi.processors.standard.ProvenanceTaskReport
```
Should namespace this to org.apache.nifi.reporting

benqiu2016 commented:
@apiri i addressed your above review comments. Adding the new test case to support full site2site is not as easy as we thought, we need to support hand shake/CRC/two phase commit to make sure that end to end works.. Right now we only support S2S client in CPP, we can add S2S server in CPP. If we can add that, it means that we can run both client and server S2S in CPP test suite. We can address that via a different JIRA.

phrocker commented:
@benqiu2016  Did you make that JIRA and make it a blocker? It should be the next most important task.

phrocker commented on libminifi/include/core/Processor.h:
```diff
@@ -239,12 +242,21 @@ class Processor : public Connectable, public ConfigurableComponent,
 
- private:
+  //! obtainSite2SiteProtocol for use
+  std::shared_ptr<Site2SiteClientProtocol> obtainSite2SiteProtocol(std::string host, uint16_t sport, uuid_t portId);
```
@benqiu2016 Can you make a ticket to move this out of Processor? It should be a separate component. If you make the ticket I'll make this particular change.

benqiu2016 replied:
Open MINIFI-269 for support site2site server protocol
Open MINIFI-270 for the refactor of above

apiri commented:
reviewing changes

apiri commented:
changes looked good.  was able to verify functionality.  licensing looks good.  will get this merged

benqiu2016 commented:
@apiri Thanks.
